lambda_code/src/models/fhir_immunization.py

```diff
@@ -754,6 +754,389 @@ def pre_validate_reason_code_coding_displays(cls, values: dict) -> dict:
+                ip_address_code,
+                generate_field_location_for_questionnnaire_response(
+                    link_id="IpAddress", field_type="code"
+                ),
+            )
+        except KeyError:
+            pass
+
+        return values
+
+    @classmethod
+    def pre_validate_user_id_code(cls, values: dict) -> dict:
+        """
+        Pre-validate that, if contained[?(@.resourceType=='QuestionnaireResponse')]
+        .item[?(@.linkId=='UserId')].answer[0].valueCoding.code (legacy CSV field name:
+        USER_ID) exists, then it is a non-empty string
+        """
+        try:
+            user_id_code = get_generic_questionnaire_response_value(
+                values, "UserId", "code"
+            )
+            pre_validate_string(
+                user_id_code,
+                generate_field_location_for_questionnnaire_response(
+                    link_id="UserId", field_type="code"
+                ),
+            )
+        except KeyError:
+            pass
+
+        return values
+
+    @classmethod
+    def pre_validate_user_name_code(cls, values: dict) -> dict:
+        """
+        Pre-validate that, if contained[?(@.resourceType=='QuestionnaireResponse')]
+        .item[?(@.linkId=='UserName')].answer[0].valueCoding.code (legacy CSV field name:
+        USER_NAME) exists, then it is a non-empty string
+        """
+        try:
+            user_name_code = get_generic_questionnaire_response_value(
+                values, "UserName", "code"
+            )
+            pre_validate_string(
+                user_name_code,
+                generate_field_location_for_questionnnaire_response(
+                    link_id="UserName", field_type="code"
+                ),
+            )
+        except KeyError:
+            pass
+
+        return values
+
+    @classmethod
+    def pre_validate_user_email_code(cls, values: dict) -> dict:
+        """
+        Pre-validate that, if contained[?(@.resourceType=='QuestionnaireResponse')]
+        .item[?(@.linkId=='UserEmail')].answer[0].valueCoding.code (legacy CSV field name:
+        USER_EMAIL) exists, then it is a non-empty string
+        """
+        try:
+            user_email_code = get_generic_questionnaire_response_value(
+                values, "UserEmail", "code"
+            )
+            pre_validate_string(
+                user_email_code,
+                generate_field_location_for_questionnnaire_response(
+                    link_id="UserEmail", field_type="code"
+                ),
+            )
+        except KeyError:
+            pass
+
+        return values
+
+    @classmethod
+    def pre_validate_submitted_time_stamp_code(cls, values: dict) -> dict:
+        """
+        Pre-validate that, if contained[?(@.resourceType=='QuestionnaireResponse')]
+        .item[?(@.linkId=='SubmittedTimeStamp')].answer[0].valueCoding.code (legacy CSV field name:
+        SUBMITTED_TIMESTAMP) exists, then it is a non-empty string
```

Comment:
Doc string should reflect checks for datetime, not just generic string (can be copied from another datetime validator e.g. occurfence_date_time)

Reply:
done